File: cli/backend/weather_data.py
import core

# TODO: Port to Rust


class WeatherData:  # TODO: Abstract more attributes
    def __init__(
        self,
        time,
        temperature: int,
        min_temp: float,
        max_temp: float,
        wind: core.WindData,
        dewpoint,
        feels_like,
        aqi: int,
        cloud_cover,
        conditions: list,
        condition_sentence,
    ):
        self.time = time
        self.temperature = temperature
        self.min_temp = min_temp
        self.max_temp = max_temp
        self.wind = wind
        self.dewpoint = dewpoint
        self.feels_like = feels_like
        self.aqi = aqi
        self.cloud_cover = cloud_cover
        self.conditions = conditions
        self.condition_sentence = condition_sentence

    def get_condition_sentence(self):
        data = self.conditions.copy()
        condition_match = data[0].sentence
        out = condition_match
        data.pop(0)
        for condition in data:
            out += ". Also, "
            condition_match = condition.sentence
            out += condition_match.lower()
        out += "."
        return out

    def get_condition_ids(self):
        ids = []
        for condition in self.conditions:
            ids.append(condition.condition_id)
        return ids

File: internal/backend/src/weather_data.rs
use serde::{Deserialize, Serialize};

use crate::WeatherCondition;
use crate::WindData;
use chrono::{DateTime, Utc, Duration};
use chrono::serde::ts_seconds;

pub fn get_conditions_sentence(conditions: Vec<WeatherCondition>) -> String {
    let mut data = conditions;
    let conditions_match = data
        .get(0)
        .expect("0th element expected")
        .sentence
        .to_string();
    let mut conditions_sentences = conditions_match;
    data.remove(0);
    for condition in data {
        conditions_sentences += ". Also, ";
        conditions_sentences += &*condition.sentence.to_lowercase();
        conditions_sentences += ".";
    }
    conditions_sentences
}

fn _default_duration() -> Duration {
    Duration::hours(1)
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub struct PrecipitationData {
    pub amount: f32,
    #[serde(skip_serializing, skip_deserializing, default = "_default_duration")] // TODO: Fix
    pub time: Duration,
    pub probability: u8,
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
pub struct WeatherData {
    #[serde(with = "ts_seconds")]
    pub time: DateTime<Utc>,
    pub temperature: f32,
    pub min_temp: f32,
    pub max_temp: f32,
    pub wind: WindData,
    pub raw_data: String,
    pub dewpoint: f32,
    pub feels_like: f32,
    pub aqi: u8,
    pub cloud_cover: u8,
    pub conditions: Vec<WeatherCondition>,
    pub condition_sentence: String,
    pub rain_data: PrecipitationData,
    pub snow_data: PrecipitationData,
}
